onnxruntime/core/providers/cpu/tensor/gather_elements.h

```diff
@@ -0,0 +1,31 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/common/common.h"
+#include "core/framework/op_kernel.h"
+#include "core/providers/common.h"
+#include "core/providers/cpu/tensor/utils.h"
+
+namespace onnxruntime {
+
+class GatherElements final : public OpKernel {
+ public:
+  GatherElements(const OpKernelInfo& info) : OpKernel(info) {
+    ORT_ENFORCE(info.GetAttr<int64_t>("axis", &axis_).IsOK(), "Missing/Invalid 'axis' attribute value");
+  }
+
+  Status Compute(OpKernelContext* context) const override;
+
+ private:
+  Status CoreImplString(const Tensor* input_tensor, const Tensor* indices_tensor,
+                        Tensor* output_tensor, int64_t axis) const;
+
+  Status CoreImpl(const Tensor* input_tensor, const Tensor* indices_tensor,
+                  Tensor* output_tensor, int64_t axis) const;
+
```

Comment:
The input, indices and output are all required, so should be references

Reply:
Hmmm - Personally I like them encapsulated within the class as private methods and prevent exposure to anything outside the kernel for now.

Reply:
How does a static file scope function get exposed elsewhere? Upside of a static method is that it can only change things in the input parameters so it reduces the amount of code that can change the internals of a class.

Reply:
well the private methods are const right ? They won't be able to change the internals of a class either... But I can make them static if you feel that is consistent with how we have been doing things. Shouldn't be a big change. But is there anything against using const private methods to do the core implementations in ?

Reply:
In general it's good to keep things that can be done in static helpers out of the class definition to keep the definition simple. In this case the helpers are implementing significant amounts of functionality so it's maybe not as clear where they should be, but I'd suggest doing it via static helpers gives you the benefits of pimpl without having to actually do pimpl. 

---
In reply to: [325460210](https://github.com/microsoft/onnxruntime/pull/1735#discussion_r325460210) [](ancestors = 325460210)

Reply:
ok - will make them static helpers

Reply:
Made them static helpers.

Reply:
Actually templatized the code a bit more. Maybe it looks better now...